File: leetcode/implement-trie-prefix-tree.cpp
// Author: Suraj Aralihalli
// Url: https://leetcode.com/problems/implement-trie-prefix-tree/
// Date: 21st November, 2021
// Tags: trie, prefix-tree

class TrieNode {
public:
    bool isWord;
    unordered_map<char, TrieNode*> children;
    TrieNode() {
        isWord = false;
        children = unordered_map<char, TrieNode*>();
    }
};

class Trie {
public:
    TrieNode* root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for(char c: word) {
            if(node->children.find(c)==node->children.end()) {
                node->children[c] = new TrieNode();
                node = node->children[c];
            }
            else {
                node = node->children[c];
            }
        }
        node->isWord = true;
        return;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        for(char c: word) {
            if(node->children.find(c)==node->children.end()) {
                return false;
            }
            else {
                node = node->children[c];
            }
        }
        return node->isWord;
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for(char c: prefix) {
            if(node->children.find(c)==node->children.end()) {
                return false;
            }
            else {
                node = node->children[c];
            }
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: leetcode/implement-trie-prefix-tree.py
# Author: Suraj Aralihalli
# Url: https://leetcode.com/problems/implement-trie-prefix-tree/
# Date: 21st November, 2021
# Tags: trie, prefix-tree


# Approach 2
class TrieNode:
    children = None
    isWord = None
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isValid = False

class Trie:
    root = None
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            node = node.children[c]
        node.isWord = True

    def search(self, word: str) -> bool:
        node = self.root
        for c in word:
            if c in node.children:
                node = node.children[c]
            else:
                return False
        if node.isWord:
            return True
        else:
            return False
    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for c in prefix:
            if c in node.children:
                node = node.children[c]
            else:
                return False
        return True
    
# Approach 1    
class Node:
    def __init__(self):
        self.children = {}
        self.isWord = False;

class Trie:

    def __init__(self):
        self.root = Node()
        
    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if(node.children.get(c)!=None):
                node = node.children[c]
            else:
                node.children[c] = Node()
                node = node.children[c]
        node.isWord = True
            

    def search(self, word: str) -> bool:
        node = self.root
        for c in word:
            if(node.children.get(c)!=None):
                node = node.children[c]
            else:
                return False
        return node.isWord
        

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for c in prefix:
            if(node.children.get(c)!=None):
                node = node.children[c]
            else:
                return False
        return True
